Range based VolumeFilter

nightshift2k opened the pull request:
## Summary
Allow VolumeFilter to extend the Volume beyond the exchanges 24h volume to a custom range in time and candlesize.

## Quick changelog

- Adds `lookback_days`, `lookback_days`, `lookback_timeframe` and `lookback_period` to pairlist config
- If  `lookback_days` or `lookback_timeframe` and `lookback_period` are specified fetches the specified amount of candles and calculates the volume based on typical price ((h+l+c)/3) in the given range
- Overwrites the `ticker['quoteVolume']` with new range based values if activated
- Behaves the usual way, if above given params are not specified

## What's new?
Improvement to VolumeFilter to calculate based on candlesize and range to allow more flexible volume pairlist.

## TODO
- *IN  PROGRESS*: Write tests to cover new cases from the new configuration possibilites
- ~~Enhance documentation for new configuration possibilities~~

nightshift2k committed:
first iteration of volume pairlist with range lookback

nightshift2k committed:
code cleanup and comments

nightshift2k committed:
fixed `self._tf_in_secs` to `self._tf_in_sec`

nightshift2k committed:
added meaningful logging of used lookback range

nightshift2k committed:
PEP8 compliance

xmatthias commented on freqtrade/plugins/pairlist/VolumePairList.py:
```diff
@@ -103,6 +141,54 @@ def filter_pairlist(self, pairlist: List[str], tickers: Dict) -> List[str]:
+                           .floor('minute')
+                           .shift(minutes=-(self._lookback_period * self._tf_in_min)
+                                  - self._tf_in_min)
+                           .float_timestamp) * 1000
```
why not use int_timestamp instead of explicitly parsing to int ... ?

nightshift2k replied:
Honestly, I stole that part exactly from RangeStabilityFilter.py so, yeah sure this makes more sense, but maybe we should also look at the other filters, to improve on that? ;)

xmatthias replied:
for sure :laughing: 
it's not like i notice everything (or the same things) every day. While i clearly should - that's now how humans work ... (at least i don't).

xmatthias commented on freqtrade/plugins/pairlist/VolumePairList.py:
```diff
@@ -36,6 +39,35 @@ def __init__(self, exchange, pairlistmanager,
+                f'Refresh period of {self._refresh_period} seconds is smaller than one '
+                f'timeframe of {self._lookback_timeframe}. Please adjust refresh_period '
+                f'to at least {self._tf_in_sec} and restart the bot.'
+            )
```
You may also wanna top-cap this at the exchange candle limit for one call (look at `AgeFilter` for how to do this).

`refresh_latest_ohlcv` will only return 1 exchange call - so you won't have 2000 5m candles available.

nightshift2k replied:
Not sure if I understand this correctly, but this is what should happen in line 84ff, where we trigger an `OperationalException` when the `lookback_period` exceedes `ohlcv_candle_limit` for given candlesize?!

Or maybe I misunderstood it.

xmatthias replied:
maybe i wrote this comment before getting to line 84 ... :laughing: 

sorry ...

xmatthias reviewed (commented):
I'd say looks great already! :+1:

nightshift2k committed:
improvements:
- `float_timestamp` switched to `int_timestamp`
- added documentation to pairlists.md

nightshift2k commented:
I will definitely need assistance further down the road to implement the appropriate tests for the new parameters. Especially how to cover all possible cases, where the historical candle data will be needed to perform the lookback, hope somone more skilled than me, can assist with that or at least hint me.

nightshift2k committed:
added `test_VolumePairList_range` to test_pairlist.py

xmatthias commented on freqtrade/plugins/pairlist/VolumePairList.py:
```diff
@@ -103,6 +141,53 @@ def filter_pairlist(self, pairlist: List[str], tickers: Dict) -> List[str]:
+                    )
+
+                    # replace quoteVolume with range quoteVolume sum calculated above
+                    filtered_tickers[i]['quoteVolume'] = pair_candles['quoteVolume'].sum()
```
Wouldn't you want to use something like `pair_candles['quoteVolume'].rolling(self._lookback_period).sum()` - and use the last value of that result?

while `since_ms` is used to specify since when we'd like to have history - i'm not certain it'll work reliably for all exchanges (depends a bit on the exchange api functioning).
if you provide a since_ms very close to the current date (closer than `now - minimum candles`  - this might result in using 500 candles even though you only specified 20 ...

nightshift2k replied:
fair point! didn't have in mind that a very tight delta between since_ms and now can result in fetching the maximum candle amount.

xmatthias replied:
doesn't even have to e very tight - some exchanges provide you a minimum of 100 candles, even if you only wanted 20 ...

nightshift2k replied:
yeah, building a total sum with a varying amount of candles depending on how the exchange behaves was stupid anyway, so this is definitely a much more safer approach 👌

nightshift2k committed:
changed quoteVolume to be built over a
rolling period using lookback_period
to avoid pair_candles being larger
than requested lookback_period

nightshift2k committed:
adjusted `test_pairlist.py` for fixed rolling sum

xmatthias commented on docs/includes/pairlists.md:
```diff
@@ -76,6 +76,35 @@ Filtering instances (not the first position in the list) will not apply any cach
+        "number_assets": 20,
+        "sort_key": "quoteVolume",
+        "refresh_period": 3600,
+        "lookback_timeframe": "1h",
+        "lookback_timeframe": 72
```
something is not right here ... 
you can't really configure lookback_timeframe twice ;)

nightshift2k replied:
🤦‍♂️ on it ....

xmatthias reviewed (commented):
I think we'll want a warning on the docs about performance of this.

I've used this with number_assets: 10 (only that filter) - and for USDT, it'll need to download data for 279 pairs - which obviously takes quite some time.

Maybe even proposing a "pre-filter" to limit the number of pairs to top 100 volume initially, and use this as a further filter (based on the assumption that the current top100 pairs will probably contain the top 50 pairs based on volume in another timeframe).

nightshift2k committed:
warning for range based lookback performance
more readable formatting of examples

xmatthias committed:
Fix typos and improve wording in docs

xmatthias reviewed (approved):
Great! :100: 

LGTM :+1:
